File: declarative_sqlite/lib/src/sync/hlc.dart
import 'package:equatable/equatable.dart';
import 'package:uuid/uuid.dart';

/// A Hybrid Logical Clock (HLC) timestamp.
///
/// HLCs combine physical time (wall-clock) with a logical counter to create
/// a partially ordered, sortable, and conflict-free timestamp across
/// distributed nodes.
///
/// The format is: `<milliseconds>:<counter>:<nodeId>`
/// - `milliseconds`: Milliseconds since the Unix epoch, left-padded with zeros.
/// - `counter`: A logical counter, left-padded with zeros, to resolve ties
///   when events occur in the same millisecond.
/// - `nodeId`: A unique identifier for the node that generated the timestamp.
///
/// The padding ensures that HLC timestamps can be lexically sorted.
class Hlc extends Equatable implements Comparable<Hlc> {
  final int milliseconds;
  final int counter;
  final String nodeId;

  @override
  List<Object?> get props =>
      [milliseconds.hashCode, counter.hashCode, nodeId.hashCode];

  @override
  bool? get stringify => false;

  static Hlc min = const Hlc(0, 0, '0');

  const Hlc(this.milliseconds, this.counter, this.nodeId);

  /// Creates an HLC timestamp from a string representation.
  factory Hlc.parse(String encoded) {
    final parts = encoded.split(':');
    if (parts.length != 3) {
      throw FormatException('Invalid HLC string format', encoded);
    }
    return Hlc(
      int.parse(parts[0]),
      int.parse(parts[1]),
      parts[2],
    );
  }

  @override
  String toString() {
    // Pad to ensure lexical sortability.
    // Milliseconds: 15 digits (supports dates far into the future)
    // Counter: 5 digits (allows for 100,000 events per millisecond)
    return '${milliseconds.toString().padLeft(15, '0')}:'
        '${counter.toString().padLeft(5, '0')}:'
        '$nodeId';
  }

  @override
  int compareTo(Hlc other) {
    if (milliseconds != other.milliseconds) {
      return milliseconds.compareTo(other.milliseconds);
    }
    if (counter != other.counter) {
      return counter.compareTo(other.counter);
    }
    return nodeId.compareTo(other.nodeId);
  }
}

/// A clock that generates Hybrid Logical Clock timestamps.
/// 
/// This is a singleton to ensure causal ordering is preserved throughout
/// the entire application. All database instances should use the same
/// HLC clock instance.
class HlcClock {
  static HlcClock? _instance;
  
  int _lastMilliseconds;
  int _lastCounter;
  final String nodeId;

  HlcClock._internal({String? nodeId})
      : _lastMilliseconds = DateTime.now().millisecondsSinceEpoch,
        _lastCounter = 0,
        nodeId = nodeId ?? const Uuid().v4();

  /// Gets the singleton instance of the HLC clock.
  /// 
  /// If [nodeId] is provided on first access, it will be used as the node ID.
  /// Subsequent calls with different node IDs will be ignored.
  factory HlcClock({String? nodeId}) {
    return _instance ??= HlcClock._internal(nodeId: nodeId);
  }

  /// Gets the singleton instance without creating it.
  /// Throws if the instance hasn't been created yet.
  static HlcClock get instance {
    if (_instance == null) {
      throw StateError('HlcClock instance not initialized. Call HlcClock() first.');
    }
    return _instance!;
  }

  /// Resets the singleton instance. Used primarily for testing.
  static void resetInstance() {
    _instance = null;
  }

  /// Generates a new HLC timestamp.
  Hlc now() {
    final wallClock = DateTime.now().millisecondsSinceEpoch;

    if (wallClock > _lastMilliseconds) {
      _lastMilliseconds = wallClock;
      _lastCounter = 0;
    } else {
      _lastCounter++;
    }

    return Hlc(_lastMilliseconds, _lastCounter, nodeId);
  }

  /// Updates the clock based on a received HLC timestamp from another node.
  /// This is crucial for keeping clocks synchronized in a distributed system.
  void update(Hlc received) {
    final wallClock = DateTime.now().millisecondsSinceEpoch;

    if (wallClock > _lastMilliseconds && wallClock > received.milliseconds) {
      // Case 1: Local and received time are both in the past.
      // The physical clock is authoritative.
      _lastMilliseconds = wallClock;
      _lastCounter = 0;
    } else if (_lastMilliseconds == received.milliseconds) {
      // Case 2: Same millisecond, update to the higher counter.
      _lastCounter =
          (received.counter > _lastCounter ? received.counter : _lastCounter) +
              1;
    } else if (_lastMilliseconds > received.milliseconds) {
      // Case 3: Local clock is ahead. Increment counter.
      _lastCounter++;
    } else {
      // Case 4: Received clock is ahead. Adopt its time.
      _lastMilliseconds = received.milliseconds;
      _lastCounter = received.counter + 1;
    }
  }
}

File: packages/core/src/sync/hlc.ts
/**
 * Hybrid Logical Clock (HLC) Implementation
 * 
 * HLC provides distributed timestamps that combine physical time with logical counters
 * for conflict-free ordering of events in distributed systems.
 * 
 * Format: `<19-digit-zero-padded-milliseconds>:<9-digit-zero-padded-counter>:<nodeId>`  
 * Example: `0000001701878400000:000000000:D5FC9B48AEB945869FD95A205CF5FBA2`
 * 
 * The zero-padding ensures that timestamps can be compared using simple string comparison,
 * which is critical for distributed systems and database indexing.
 */

export interface HlcTimestamp {
  milliseconds: number;
  counter: number;
  nodeId: string;
}

export class Hlc {
  private lastMilliseconds: number = 0;
  private counter: number = 0;

  constructor(private nodeId: string) {
    if (!nodeId || nodeId.length === 0) {
      throw new Error('HLC requires a non-empty nodeId');
    }
  }

  /**
   * Generate a new HLC timestamp
   */
  now(): HlcTimestamp {
    const physicalTime = Date.now();

    if (physicalTime > this.lastMilliseconds) {
      // Physical time advanced, reset counter
      this.lastMilliseconds = physicalTime;
      this.counter = 0;
    } else {
      // Physical time hasn't advanced, increment counter
      this.counter++;
    }

    return this.createTimestamp(
      this.lastMilliseconds,
      this.counter,
      this.nodeId
    );
  }

  /**
   * Update HLC based on received timestamp (for clock synchronization)
   */
  update(received: HlcTimestamp): HlcTimestamp {
    const physicalTime = Date.now();
    const maxMilliseconds = Math.max(physicalTime, this.lastMilliseconds, received.milliseconds);

    if (maxMilliseconds === this.lastMilliseconds && maxMilliseconds === received.milliseconds) {
      // Both clocks at same time, use max counter + 1
      this.counter = Math.max(this.counter, received.counter) + 1;
    } else if (maxMilliseconds === this.lastMilliseconds) {
      // Our clock is ahead
      this.counter++;
    } else if (maxMilliseconds === received.milliseconds) {
      // Received clock is ahead
      this.counter = received.counter + 1;
    } else {
      // Physical time advanced beyond both
      this.counter = 0;
    }

    this.lastMilliseconds = maxMilliseconds;

    return this.createTimestamp(
      this.lastMilliseconds,
      this.counter,
      this.nodeId
    );
  }

  /**
   * Create an HLC timestamp with automatic serialization support
   */
  private createTimestamp(
    milliseconds: number,
    counter: number,
    nodeId: string
  ): HlcTimestamp {
    const timestamp: HlcTimestamp = {
      milliseconds,
      counter,
      nodeId,
    };

    // Add toString() method for automatic serialization
    Object.defineProperty(timestamp, 'toString', {
      value: () => Hlc.toString(timestamp),
      enumerable: false,
    });

    return timestamp;
  }

  /**
   * Serialize HLC timestamp to string
   * Format: 19-digit milliseconds : 9-digit counter : nodeId
   */
  static toString(timestamp: HlcTimestamp): string {
    const milliseconds = timestamp.milliseconds.toString().padStart(19, '0');
    const counter = timestamp.counter.toString().padStart(9, '0');
    return `${milliseconds}:${counter}:${timestamp.nodeId}`;
  }

  /**
   * Parse HLC timestamp from string
   */
  static parse(hlcString: string): HlcTimestamp {
    const parts = hlcString.split(':');
    if (parts.length !== 3) {
      throw new Error(`Invalid HLC format: ${hlcString}. Expected format: 0000000000000000000:000000000:NODEID`);
    }

    const milliseconds = parseInt(parts[0]!, 10);
    const counter = parseInt(parts[1]!, 10);
    const nodeId = parts[2]!;

    if (isNaN(milliseconds) || isNaN(counter)) {
      throw new Error(`Invalid HLC format: ${hlcString}. milliseconds and counter must be numbers`);
    }

    const timestamp: HlcTimestamp = { milliseconds, counter, nodeId };

    // Add toString() method for automatic serialization
    Object.defineProperty(timestamp, 'toString', {
      value: () => Hlc.toString(timestamp),
      enumerable: false,
    });

    return timestamp;
  }

  /**
   * Compare two HLC timestamps
   * Returns: -1 if a < b, 0 if a === b, 1 if a > b
   */
  static compare(a: HlcTimestamp, b: HlcTimestamp): number {
    // First compare milliseconds
    if (a.milliseconds < b.milliseconds) return -1;
    if (a.milliseconds > b.milliseconds) return 1;

    // Then compare counter
    if (a.counter < b.counter) return -1;
    if (a.counter > b.counter) return 1;

    // Then compare nodeId (for deterministic ordering)
    if (a.nodeId < b.nodeId) return -1;
    if (a.nodeId > b.nodeId) return 1;

    return 0;
  }

  /**
   * Check if timestamp a is before timestamp b
   */
  static isBefore(a: HlcTimestamp, b: HlcTimestamp): boolean {
    return this.compare(a, b) < 0;
  }

  /**
   * Check if timestamp a is after timestamp b
   */
  static isAfter(a: HlcTimestamp, b: HlcTimestamp): boolean {
    return this.compare(a, b) > 0;
  }

  /**
   * Get the maximum of two timestamps
   */
  static max(a: HlcTimestamp, b: HlcTimestamp): HlcTimestamp {
    return this.compare(a, b) >= 0 ? a : b;
  }
}
